swarmcg/swarmCG.py

```diff
@@ -1222,42 +1197,124 @@ def get_initial_guess_list(ns, nb_particles):
+				)  # no need for PBC handling, trajectories were made wholes for the molecule
+			coord[:, bead_id, :] = traj
+
+	ns.aa2cg_universe.load_new(coord, format=mda.coordinates.memory.MemoryReader)
+
+	# virtual sites are mapped using previously defined regular beads positions and appropriate virtual sites functions
+	# it is also possible to use a VS for defining another VS position, if the VS used for definition is defined before
+	# no need to check if the functions used for VS definition are correct here, this has been done already
+	for bead_id in range(len(ns.cg_itp['atoms'])):
+		if ns.cg_itp['atoms'][bead_id]['bead_type'].startswith('v'):
+
+			traj = np.empty((len(ns.aa2cg_universe.trajectory), 3))
+
+			if ns.cg_itp['atoms'][bead_id]['vs_type'] == 2:
+				vs_def_beads_ids = ns.cg_itp['virtual_sites2'][bead_id]['vs_def_beads_ids']
+				vs_params = ns.cg_itp['virtual_sites2'][bead_id]['vs_params']
+
+				if ns.cg_itp['virtual_sites2'][bead_id]['func'] == 1:
+					vsf.vs2_func_1(ns, traj, vs_def_beads_ids, vs_params, bead_id)
+				elif ns.cg_itp['virtual_sites2'][bead_id]['func'] == 2:
+					vsf.vs2_func_2(ns, traj, vs_def_beads_ids, vs_params, bead_id)
+
+			if ns.cg_itp['atoms'][bead_id]['vs_type'] == 3:
+				vs_def_beads_ids = ns.cg_itp['virtual_sites3'][bead_id]['vs_def_beads_ids']
+				vs_params = ns.cg_itp['virtual_sites3'][bead_id]['vs_params']
 
-	return
+				if ns.cg_itp['virtual_sites3'][bead_id]['func'] == 1:
+					vsf.vs3_func_1(ns, traj, vs_def_beads_ids, vs_params)
+				elif ns.cg_itp['virtual_sites3'][bead_id]['func'] == 2:
+					vsf.vs3_func_2(ns, traj, vs_def_beads_ids, vs_params, bead_id)
+				elif ns.cg_itp['virtual_sites3'][bead_id]['func'] == 3:
+					vsf.vs3_func_3(ns, traj, vs_def_beads_ids, vs_params)
+				elif ns.cg_itp['virtual_sites3'][bead_id]['func'] == 4:
+					vsf.vs3_func_4(ns, traj, vs_def_beads_ids, vs_params)
+
+			# here it's normal there is only function 2, that's the only one that exists in gromacs for some reason
+			if ns.cg_itp['atoms'][bead_id]['vs_type'] == 4:
+				vs_def_beads_ids = ns.cg_itp['virtual_sites4'][bead_id]['vs_def_beads_ids']
+				vs_params = ns.cg_itp['virtual_sites4'][bead_id]['vs_params']
+
+				if ns.cg_itp['virtual_sites4'][bead_id]['func'] == 2:
+					vsf.vs4_func_2(ns, traj, vs_def_beads_ids, vs_params)
+
+			if ns.cg_itp['atoms'][bead_id]['vs_type'] == 'n':
+				vs_def_beads_ids = ns.cg_itp['virtual_sitesn'][bead_id]['vs_def_beads_ids']
+				vs_params = ns.cg_itp['virtual_sitesn'][bead_id]['vs_params']
+
+				if ns.cg_itp['virtual_sitesn'][bead_id]['func'] == 1:
+					vsf.vsn_func_1(ns, traj, vs_def_beads_ids)
+				elif ns.cg_itp['virtual_sitesn'][bead_id]['func'] == 2:
+					vsf.vsn_func_2(ns, traj, vs_def_beads_ids, bead_id)
+				elif ns.cg_itp['virtual_sitesn'][bead_id]['func'] == 3:
+					vsf.vsn_func_3(ns, traj, vs_def_beads_ids, vs_params)
+
+			coord[:, bead_id, :] = traj
+
+	ns.aa2cg_universe.load_new(coord, format=mda.coordinates.memory.MemoryReader)
```

Comment:
Mapping the trajectory like this is a whole lot cleaner and faster than what we were doing previously